File: NEUOJ-搜索/分组序列.cpp
#include <stdio.h>
#include <stdlib.h>
#define int long long
int check(int w[], int n, int k, int P) {
    int cnt = 1; // 至少一组
    int sum = 0;
    for (int i = 0; i < n; i++) {
        if (sum + w[i] > P) {
            cnt++;
            sum = w[i];
        } else {
            sum += w[i];
        }
        if (cnt > k) return 0; // 超过k组，不可行
    }
    return cnt <= k; // 检查是否满足分组数
}

signed main() {
    int n, k;
    scanf("%lld %lld", &n, &k);
    int *w = (int *)malloc(sizeof(int) * n);
    int max_val = 0, sum_val = 0;
    for (int i = 0; i < n; i++) {
        scanf("%lld", &w[i]);
        if (w[i] > max_val) max_val = w[i];
        sum_val += w[i];
    }
    int left = max_val, right = sum_val;
    while (left < right) { // 二分查找
        int mid = (left + right) / 2;
        if (check(w, n, k, mid)) {
            right = mid; // 可行，尝试更小值
        } else {
            left = mid + 1; // 不可行，增大下界
        }
    }
    printf("%lld\n", left);
    free(w);
    return 0;
}

File: NEUOJ-搜索/分组序列.py
n, k = map(int, input().split())
x = list(map(int, input().split()))
max_x = max(x)
sum_x = sum(x)

left = max_x
right = sum_x

def is_possible(mid):
    if mid < max_x:
        return False
    count = 1
    current_sum = 0
    for num in x:
        if current_sum + num > mid:
            count += 1
            current_sum = num
            if count > k:
                return False
        else:
            current_sum += num
    return True

while left < right:
    mid = (left + right) // 2
    if is_possible(mid):
        right = mid
    else:
        left = mid + 1

print(left)